### PR TITLE
Fix compound math

### PR DESCRIPTION
**Pull request by MerlinEgalite**

### CODE REVIEW

File: test/TestCompoundMath.sol [hunk 1]
```diff
@@ -18,17 +19,25 @@ contract TestCompoundMath is Test {
...
 
-    function testDiv(uint128 _x, uint128 _y) public {
-        vm.assume(_y > 0);
+    function testDiv(uint256 x, uint256 y) public {
+        unchecked {
+            if (y == 0 || (x != 0 && (x * SCALE) / x != SCALE)) {
+                vm.expectRevert();
+                CompoundMath.div(x, y);
+            }
+        }
```
Comment (Rubilmax) [inline test/TestCompoundMath.sol:39]: same here

**Review by Rubilmax — APPROVED** [review]

File: src/math/CompoundMath.sol [hunk 2]
```diff
@@ -15,13 +15,25 @@ library CompoundMath {
...
+
+            z := div(z, WAD)
         }
     }
 
     function div(uint256 x, uint256 y) internal pure returns (uint256 z) {
         assembly {
-            z := div(div(mul(x, SCALE), WAD), y)
+            z := mul(x, SCALE)
+            // Revert if y > 0 or x > 0 and (x * SCALE) / x != SCALE
```
Comment (QGarchery) [inline src/math/CompoundMath.sol:31]:
```suggestion
            // Revert when y > 0 or when both x > 0 and (x * SCALE) / x != SCALE
```

File: test/TestCompoundMath.sol [hunk 3]
```diff
@@ -18,20 +19,40 @@ contract TestCompoundMath is Test {
 
     /// TESTS ///
 
-    function testMul(uint128 _x, uint128 _y) public {
-        uint256 x = _x;
-        uint256 y = _y;
+    function testMul(uint256 x, uint256 y) public {
+        unchecked {
+            vm.assume(y == 0 || (y > 0 && (x * y) / y == x));
+        }
```
Comment (MathisGD) [inline test/TestCompoundMath.sol:25]: Not sure why you have the y==0 ?

File: src/math/CompoundMath.sol [hunk 4]
```diff
@@ -15,13 +15,25 @@ library CompoundMath {
...
+            if or(iszero(y), iszero(or(iszero(x), eq(div(z, x), SCALE)))) {
+                revert(0, 0)
```
Comment (QGarchery) [inline src/math/CompoundMath.sol:33]: Would this work ?
```suggestion
            if or(iszero(y), and(x, lt(div(z, x), SCALE))) {
                revert(0, 0)
```

**Review by QGarchery — APPROVED** [review]

File: test/TestCompoundMath.sol [hunk 5]
```diff
@@ -18,20 +19,40 @@ contract TestCompoundMath is Test {
...
     }
 
+    function testDivOverflow(uint256 x, uint256 y) public {
+        unchecked {
+            vm.assume(y == 0 || (x != 0 && (x * SCALE) / x != SCALE));
+        }
+
+        vm.expectRevert();
+        CompoundMath.div(x, y);
+    }
```
Comment (MathisGD) [inline test/TestCompoundMath.sol:54]: I would have separate the case overflow and the case div by 0

File: src/math/CompoundMath.sol (hunk 4 above)
Comment (QGarchery) [reply src/math/CompoundMath.sol:33]: Actually no, because `or` is a bitwise `and`

**Review by MathisGD — APPROVED** [review]

File: src/math/CompoundMath.sol (hunk 4 above)
Comment (QGarchery) [inline src/math/CompoundMath.sol:33]:
```suggestion
            if or(iszero(y), and (gt(x, 0), lt(div(z, x), SCALE))) {
                revert(0, 0)
```

**Pull request merged**